### PR TITLE
Add port location information to verilog output

### PR DESCRIPTION
**Pull request by youngar**
The goal of this PR is to parse in FIRRTL debug source locations for module ports, and maintain them through the pipeline, eventually printing the locations as part of the port list.  To accomplish this whole pipeline support, changes were required to the FIRRTL and HW dialects.  The changes to HW modules motivated changes to the MSFT dialect to ensure that the generic module functionality still worked for both dialects.  The port locations give better error messages when emitting a diagnostic on a block argument value.

The first design point is that I decided to support port locations on FIRRTL `FExtModuleOp`s.  With regular modules, we can attach port locations to all the block arguments.  For `FExtModuleOp`s we require a `portLocations` attribute of type `ArrayAttr<LocationAttr>` to track the port locations. To keep things simple, I decided to add the `portLocation` attribute to all kinds of FIRRTL modules.  For regular `FModuleOp`s, we copy these locations into the block argument values and add a verifier to ensure that these are kept in sync.

The FIRRTL parser was storing the Module's location in each of the block arguments, despite having code for parsing port specific location information.  It was modified to stop overriding the port's parsed location.

Port location information was added to the `FModuleLike` op interface.

Some internal FIRRTL helpers were refactored to have only common functionality, with `FModuleOp` handling the specifics of owning a region after calling the generic helpers.

The FIRRTL MLIR module printer was modified to print the debug locations when the `--mlir-print-debuginfo` option is specified.  There appears to be a problem in the printer's support for debuginfo, and it will create attribute aliases for all locations even if it does not end up printing the location.  To get around this, we query the OpPrinterFlags directly, which will be populated by the command line options.  The downside of this approach is that if someone is programmatically modifying the printer's OpPrinterOptions, we will not respect those settings. I don't think this will be an issue in practice, but it would be nice to fix this upstream.

The FIRRTL IR module parser was modified to read in an optional source location when parsing the port list.  If no location is found, it uses a location pointing to the MLIR file.

Dedup was modified to properly combine the location information of modules and their block arguments.

LowerTypes was modified to spread the port location information to each lowered port.

LowerToHW was modified to properly transfer the location information during lowering.

In the HW dialect, I continued the design point to allow port location information to be tracked by `hw.module.extern`, which required adding two attributes for tracking locations; an `argLocs` and a `resultLocs` attribute was added to HW modules.

Similar to the FIRRTL dialect many HW module helpers were updated to support these new attributes.

The HW pass FlattenIO was modified to spread the locations to flattened ports the same way as LowerTypes.

ExportVerilog was updated to allow the printing of location attributes directly, instead of taking a set of operations.  Some printer bugs were fixed here especially around consecutive 0-width ports, and some better alignment.  See `test/Conversion/ExportVerilog/port-decl-sharing.mlir` for some tests. @dtzSiFive Will you please take a look at these changes, I see sometimes we are using nbsp and other times space literals and I don't know the difference.

ExportVerilog now always prints the closing `);` on a newline when there is a non-empty port-list.  This was needed to support trailing location comments.  Technically, we could still print it on the same line when location information is not emitted for the last port in the list, but I didn't consider it worth the effort. As an example:
```verilog
// Was:
module Foo(
  output [3:0] if_0);

// Now:
module Foo(
  output [3:0] if_0 // foo.scala:10:20
);
``` 

The MSFT and SystemC dialects has to be updated to support these new attributes.  The MSFT module was updated to support the same form of location attributes as HW modules.  SystemC modules are, for now, using the common functionality but throwing away the attributes. This way, we could maintain a common set of code for creating, parsing, printing, updating modules.

The ESI ESIPortsPass was updated to spread the locations of `ChannelTypes` to the ready/valid ports it lowers them to.

The python bindings were updated to attach these new attributes to `HWModuleOp`s.  These location always default to `unkown`, for the time being. Since `LocationAttr`s are not represented in the upstream MLIR API, we had to create the attribute by calling out to the MLIR parser.

### CODE REVIEW

File: include/circt/Dialect/MSFT/MSFTOps.td [hunk 1]
```diff
@@ -133,7 +139,10 @@ def MSFTModuleOp : MSFTModuleOpBase<"module",
       TypeAttrOf<FunctionType>:$function_type,
       OptionalAttr<DictArrayAttr>:$arg_attrs,
       OptionalAttr<DictArrayAttr>:$res_attrs,
-      StrArrayAttr:$argNames, StrArrayAttr:$resultNames,
+      StrArrayAttr:$argNames,
+      StrArrayAttr:$resultNames,
+      LocationArrayAttr:$argLocs,
+      LocationArrayAttr:$resultLocs,
```
Comment (youngar) [inline include/circt/Dialect/MSFT/MSFTOps.td:140]: @teqdruid This required a bunch of changes to the ESI / MSFT modules, are you ok with adding these attributes to MSFTModules?

File: lib/Bindings/Python/circt/dialects/_hw_ops_ext.py [hunk 2]
```diff
@@ -146,17 +146,23 @@ def __init__(
 
     input_types = []
     input_names = []
+    input_locs = []
     for (i, (port_name, port_type)) in enumerate(input_ports):
       input_types.append(port_type)
       input_names.append(StringAttr.get(str(port_name)))
+      input_locs.append(Attribute.parse("loc(unknown)"))
```
Comment (youngar) [inline lib/Bindings/Python/circt/dialects/_hw_ops_ext.py:153]: @teqdruid They python bindings had to be updated to attach these new attributes.  Right now they are defaulting to `unknown`.  If a block is attached to the HWModuleOp, the locations stored in the block arguments should match these attributes. If we want better location information, here is where we would add it.

Also as a side note, we needed a `LocationAttr` not a `Location` here, so we had to reach for `Attribute.parse`.

**Comment by teqdruid** [comment]
This sounds great and I fully support the changes to ESI and MSFT at a high level. I'll do a detailed review of those changes soon, but don't let my review gate merging this PR.

File: include/circt/IR/CIRCTAttributes.td [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+#ifndef CIRCT_IR_CIRCTATTRIBUTES_TD
+#define CIRCT_IR_CIRCTATTRIBUTES_TD
+
+include "mlir/IR/OpBase.td"
+
+def LocationAttr : Attr<CPred<"$_self.isa<::mlir::LocationAttr>()">,
+  "location attribute">;
+
+def LocationArrayAttr : TypedArrayAttrBase<LocationAttr,
```
Comment (teqdruid) [inline include/circt/IR/CIRCTAttributes.td:21]: This seems like it should be upstream.

File: lib/Bindings/Python/circt/dialects/_hw_ops_ext.py (hunk 2 above)
Comment (teqdruid) [reply lib/Bindings/Python/circt/dialects/_hw_ops_ext.py:153]: It `LocationAttr` is missing from `mlir.ir`, it should be added there.

**Comment by teqdruid** [comment]
Looking at the test changes, this PR doesn't seem to change the syntax of `hw.module` so the locationattrs would have to be part of the attr-dict, yes? If so, that's not exactly ideal. Similarly (based entirely on the test changes), it looks like ExportVerilog prints `//` regardless of whether or not the location of the port is known. Is that consistent with the other location output?

**Comment by youngar** [comment]
@teqdruid The `//` in those tests is followed by a proper location, I just wasn't matching it.  The real issue here is the line `CHECK-NEXT: c` yielded the a FileCheck error along the lines of `"c" was matched on the previous line`, so I matched the comment character to help the FileCheck disambiguate the lines. (As a side note, you would think that the `-NEXT` modifier would cause FileCheck to jump to the next line of input).

In `ExportVerilog`, when the location is `UnknownLoc`, then no Verilog comment character is printed.  This is handled by the `emitLocationInfo` function.

The syntax of HWModule was changed, and I should add some dedicated tests for this.  It prints the port location information in the portlist, and ellides the attribute from the `attr-dict`, big example of all the parts of a hw module:
 `(%a : !hw.int<3> [{a = "argAttr}] loc("file.py":1:1)) -> (b : !hw.int<3> [{b = "argAttr"}] loc("out.py":1:1))`

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 4]
```diff
@@ -435,16 +435,10 @@ static void collectFileLineColLocs(Location loc,
 
 /// Return the location information as a (potentially empty) string.
 static std::string
-getLocationInfoAsStringImpl(const SmallPtrSetImpl<Operation *> &ops) {
+getLocationInfoAsStringImpl(const SmallPtrSet<Attribute, 8> &locationSet) {
```
Comment (teqdruid) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:438]: nit: Plz continue to use `SmallPtrSetImpl`.

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 5]
```diff
@@ -466,13 +460,13 @@ getLocationInfoAsStringImpl(const SmallPtrSetImpl<Operation *> &ops) {
   }
 
   // Sort the entries.
-  SmallVector<FileLineColLoc, 8> locVector;
```
Comment (teqdruid) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:469]: nit: Why did you rename this? Just creates a whole lot more code churn.

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 6]
```diff
@@ -525,29 +519,37 @@ getLocationInfoAsStringImpl(const SmallPtrSetImpl<Operation *> &ops) {
   return sstr.str();
 }
 
-/// Return the location information in the specified style.
 static std::string
-getLocationInfoAsString(const SmallPtrSetImpl<Operation *> &ops,
-                        LoweringOptions::LocationInfoStyle style) {
+getLocationInfoAsStringImpl(const SmallPtrSet<Attribute, 8> &locationSet,
```
Comment (teqdruid) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:523]: nit: Prefer `Impl` variants for outputs.

File: lib/Dialect/ESI/ESIPasses.cpp [hunk 7]
```diff
@@ -509,21 +516,25 @@ bool ESIPortsPass::updateFunc(HWModuleOp mod) {
...
+    newArgLocs.push_back(oldResultLoc);
     updated = true;
   }
 
   // Append the ready list signals we remembered above.
   for (const auto &readySig : newReadySignals) {
     newResultTypes.push_back(i1);
-    newResultNames.push_back(readySig.second);
-    newOutputOperands.push_back(readySig.first);
+    newOutputOperands.push_back(std::get<0>(readySig));
```
Comment (teqdruid) [inline lib/Dialect/ESI/ESIPasses.cpp:535]: super nit: I generally prefer unboxing tuples in the loop. We upgraded the c++ version (IIRC since this code was written) to a version which supports unboxing.

**Review by teqdruid — COMMENTED** [review]
The ESI and MSFT changes look good to me. I appreciate that you've moved some more functionality into the `hw::module_like_impl` shared code. I left a few nit comments elsewhere while doing a superficial skim.

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp (hunk 4 above)
Comment (youngar) [reply lib/Conversion/ExportVerilog/ExportVerilog.cpp:438]: 🤦  thanks

File: lib/Dialect/ESI/ESIPasses.cpp (hunk 7 above)
Comment (youngar) [reply lib/Dialect/ESI/ESIPasses.cpp:535]: Its much better that way 👍

**Comment by youngar** [comment]
Some more updates, not sure why I didn't just add a better commit message..
- Added a verifier to the MSFTModuleOp that made sure that the location attributes matched the block argument attributes
- Fixed HW pass FlattenIO to keep the restore the block argument locations after signature conversion deletes them (*sigh*)
- HWArithToHWPass restores the block argument locations after SignatureConversion deletes them
- ESI LowerPorts pass changes had a small bug, using an unknown loc for the block argument instead of the proper location
- MSFTPartition pass now preserves the location information
- HW pass StripDebugInfoWithPred now strips location information from the port location attributes

File: lib/Dialect/HW/HWOps.cpp [hunk 8]
```diff
@@ -747,8 +793,14 @@ void HWModuleOp::build(OpBuilder &builder, OperationState &result,
...
-  for (auto elt : ports.inputs)
-    body->addArgument(elt.type, builder.getUnknownLoc());
+  auto unknownLoc = builder.getUnknownLoc();
+  for (auto port : ports.inputs) {
+    auto loc = port.loc ? Location(port.loc) : unknownLoc;
+    auto type = port.type;
+    if (port.isInOut() && !type.isa<InOutType>())
+      type = InOutType::get(type);
+    body->addArgument(type, loc);
+  }
```
Comment (uenoku) [inline lib/Dialect/HW/HWOps.cpp:803]: Could you explain why this change was necessary?

Comment (youngar) [reply lib/Dialect/HW/HWOps.cpp:803]: The first is more obvious, the block argument should be using the port location specified in the PortInfo.  In the HW dialect the LocationAttr in the PortInfo structure is nullable (does not default to UnknownLoc), so I handle that case here. In the FIRRTL dialect's PortInfo, the Location is not nullable, which I think is better, but it would make this change a lot more intrusive to do the same in the HW dialect.

The second thing is that I added a verifier that ensured the block argument locations of a module match the ArgLocs attribute on the module.  While I was there I added a verifier that the function type of the module matched the block argument types, and discovered that this verifier actually fails. It turns out when we create the PortInfo object if the port is a HW::InOutType,[ we strip it and set the isInOut field to true](https://github.com/llvm/circt/blob/main/lib/Dialect/HW/HWOps.cpp#L886).  When we read in the PortInfo object [we respect this field while constructing the function type](https://github.com/llvm/circt/blob/main/lib/Dialect/HW/HWOps.cpp#L559), but were not respecting the `isInOut` [when creating the block arguments](https://github.com/llvm/circt/blob/main/lib/Dialect/HW/HWOps.cpp#L751). This is a pretty bad bug and could be moved out into its own change, I just haven't mustered the energy to do it yet.

Comment (youngar) [reply lib/Dialect/HW/HWOps.cpp:803]: Moved out into this change: https://github.com/llvm/circt/pull/4583

File: include/circt/IR/CIRCTAttributes.td (hunk 3 above)
Comment (youngar) [reply include/circt/IR/CIRCTAttributes.td:21]: upstreamed here: https://reviews.llvm.org/D142148

File: lib/Bindings/Python/circt/dialects/_hw_ops_ext.py (hunk 2 above)
Comment (youngar) [reply lib/Bindings/Python/circt/dialects/_hw_ops_ext.py]: Python bindings: https://reviews.llvm.org/D142522

**Comment by youngar** [comment]
OK I have rebased this on main, moved a few things out in to separate PRs, and switched to using some upstreamed utilities.  I plan to give this a once over and then merge it.
